src/headless/yare_impl.rs

```diff
@@ -258,6 +277,73 @@ impl<F: Fn(u32)> Headless<F> {
         }
 
         // TODO do game logic
+
+        let mut charging_spirits = Vec::new();
+        // first update stars
+        for star in &mut self.stars {
+            charging_spirits.push(Vec::<usize>::new());
+            if star.energy < star.energy_cap {
+                let gain = (star.energy as f32 / 100.).round();
+                let new_energy = star.energy + 3 + gain as i32;
+                star.energy = new_energy.min(star.energy_cap);
+            }
+        }
+
+        // process energize/explode commands + outposts
+        for (player_i, player_commands) in all_commands.iter().enumerate() {
+            let player = &mut self.players[player_i];
+
+            // TODO: I think we need to track the change in each sprites energy that tick
+            // and apply it after all of the commands process
+            for command in player_commands.iter() {
+                match command {
+                    Command::Energize{index, target} => {
+                        let source_spirit = self.spirits[*index];
```

Comment:
We should check that the spirit which is being commanded is owned by the player.